velox/docs/develop/aggregate-functions.rst

```diff
@@ -109,25 +109,323 @@ accordingly.  Data is stored in the following order:
+    static constexpr bool default_null_behavior_ = false;
+
+    // Optional.
+    static bool toIntermediate(
+      exec::out_type<Array<Generic<T1>>>& out,
+      exec::optional_arg_type<Generic<T1>> in);
+
+    struct AccumulatorType { ... };
+  };
+
+The author declares the function's input type, intermediate type, and output
+type in the simple aggregation function class. The input type must be the
+function's argument type(s) wrapped in a Row<> even if the function only takes
+one argument.
+
+The author can define an optional flag `default_null_behavior_` indicating
+whether the aggregation function has default-null behavior. This flag is true
+by default. Next, the class can have an optional method `toIntermediate()`
+that converts the aggregation function's raw input directly to its intermediate
+states. Finally, the author must define a struct named `AccumulatorType` in
+the aggregation function class. We explain each part in more details below.
+
+Default-Null Behavior
+^^^^^^^^^^^^^^^^^^^^^
+
+When adding raw inputs or intermediate states to accumulators, aggregation
+functions of default-null behavior ignore the input values that are nulls. For
+raw inputs that consist of multiple columns, an entire row is ignored if at
+least one column is null at this row. Below is an example.
+
+.. code-block:: sql
+
+  SELECT sum(c0) FROM (values (cast(NULL as bigint), 10), (NULL, 20), (NULL, 30)) AS t(c0, c1); -- NULL
+
+When generating intermediate or final output results from accumulators,
+aggregation functions of default-null behavior produce nulls for groups of no
+input row or only null rows. Another example is given below.
+
+.. code-block:: sql
+
+  SELECT sum(c0) FROM (values (1, 10), (2, 20), (3, 30)) AS t(c0, c1) WHERE c1 > 40; -- NULL
+
+Most aggregation functions have default-null behavior. An example is in
+SimpleAverageAggregate.cpp. On the other hand, SimpleArrayAggAggregate.cpp has
+an example of non-default-null behavior.
+
+This flag affects the C++ function signatures of `toIntermediate()` and methods
+in the `AccumulatorType` struct.
+
+toIntermediate
+^^^^^^^^^^^^^^
+
+The author can optionally define a static method `toIntermediate()` that
+converts a raw input to an intermediate state. If defined, this function is
+used in `abandonedPartialAggregation`. If the aggregaiton function has
+default-null behavior, the toIntermediate() function has an out-parameter
+of the type `exec::out_type<IntermediateType>&` followed by in-parameters of
+the type `exec::arg_type<T>` for each `T` wrapped inside InputType . If the
+aggregation function has non-default null behavior, the in-parameters of
+toIntermediate() are of the type `exec::optional_arg_type<T>` instead.
+
+When `T` is a primitive type except Varchar and Varbinary, `exec::arg_type<T>`
+is simply `T` itself and `exec::out_type<T>` is `T&`. `exec::optional_arg_type<T>`
+is `std::optional<T>`.
+
+When `T` is Varchar, Varbinary, or a complex type, `exec::arg_type<T>`,
```

Comment:
Can T be some opaque type ?

Reply:
The simple aggregation function interface uses the same set of vector readers and vector writers as the simple scalar interface, so yes, T can be opaque types, e.g., exec::arg_type<std::shared_ptr<CustomType>>.